keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,348 @@
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place
+- [ ] (R) Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [etcd-io/website], for publication to [etcd.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[etcd.io]: https://etcd.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[etcd-io/website]: https://github.com/etcd-io/website
+
+## Summary
+
+We are introducing functionalities to determine if a feature is enabled for the whole etcd cluster. Users would be able to turn features on or off and query cluster feature enablement in a consistent way on the cluster level. 
+
+## Motivation
+
+In [KEP-4578](https://github.com/kubernetes/enhancements/issues/4578) we introduced the feature gate framework into etcd. But the feature gate only works on the local server level. There are various features that require all members working consistently together, so we need to add functionalities to the feature gate to work on the cluster level for these features. 
+
+### Goals
+
+* Establish a mechanism to safely and consistently enable/disable cluster-level features
+
+* Introduce apis/methods to query the enablement state of a cluster level feature.
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+It would be easier for developers to add new features that might affect raft or etcd consensus without worrying too much about breaking etcd , because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate on the cluster level.
+
+The developer also would not need to worry about feature compatibility when there is a mixed version cluster.
+
+#### Story 2
+
+Users of etcd cluster would be able to find out which features are enabled in the etcd cluster and decide on how to use them downstream. 
+
+Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.
```

Comment:
`Kubernetes` means `kube-apiserver` right? (just to see if there's any other clients on k8s that reads from etcd)

Reply:
correct.